public/js/components/Sources.js

```diff
@@ -3,68 +3,261 @@
+      subpaths.splice(where, 0, pathItem);
+      subtree = subpaths[where];
+    }
+
+    // Keep track of the subpaths so we can tag each node with them.
+    path = path + "/" + part;
+  }
+
+  // Store the soure in the final created node.
+  if (isDir) {
+    setNodeContents(
+      subtree,
+      [createNode("(index)", source.get("url"), source)]
+    );
+  } else {
+    setNodeContents(subtree, source);
+  }
+}
+
+// This is inline because it's much faster. We need to revisit how we
+// load SVGs, at least for components that render them several times.
+let Arrow = (props) => {
+  return dom.span(
+    props,
+    dom.svg(
+      { viewBox: "0 0 16 16" },
+      dom.path({ d: "M8 13.4c-.5 0-.9-.2-1.2-.6L.4 5.2C0 4.7-.1 4.3.2 3.7S1 3 1.6 3h12.8c.6 0 1.2.1 1.4.7.3.6.2 1.1-.2 1.6l-6.4 7.6c-.3.4-.7.5-1.2.5z" }) // eslint-disable-line max-len
     )
   );
+};
+Arrow = React.createFactory(Arrow);
+
+let SourcesTree = React.createClass({
+  propTypes: {
+    sources: ImPropTypes.map.isRequired,
+    selectSource: PropTypes.func.isRequired
+  },
+
+  makeInitialState(props) {
+    const tree = createNode("root", "", []);
+    for (let source of props.sources.valueSeq()) {
+      addToTree(tree, source);
+    }
+
+    return { sourceTree: tree,
+             parentMap: createParentMap(tree),
+             focusedItem: null };
+  },
+
+  getInitialState() {
+    return this.makeInitialState(this.props);
+  },
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.sources !== this.props.sources) {
+      if (nextProps.sources.size === 0) {
+        this.setState(this.makeInitialState(nextProps));
+        return;
+      }
+
+      const next = Set(nextProps.sources.valueSeq());
+      const prev = Set(this.props.sources.valueSeq());
+      const newSet = next.subtract(prev);
+
+      const tree = this.state.sourceTree;
+      for (let source of newSet) {
+        addToTree(tree, source);
+      }
+
+      this.setState({ sourceTree: tree,
+                      parentMap: createParentMap(tree) });
+    }
+  },
+
+  focusItem(item) {
+    this.setState({ focusedItem: item });
+  },
+
+  selectItem(item) {
+    if (!nodeHasChildren(item)) {
+      this.props.selectSource(nodeContents(item).toJS());
+    }
+  },
+
+  render() {
+    const { focusedItem, sourceTree, parentMap } = this.state;
+
+    const tree = ManagedTree({
+      getParent: item => {
+        return parentMap.get(item);
+      },
+      getChildren: item => {
+        if (nodeHasChildren(item)) {
+          return nodeContents(item);
+        }
+        return [];
+      },
+      getRoots: () => nodeContents(sourceTree),
+      getKey: (item, i) => nodePath(item),
+      itemHeight: 30,
+      autoExpandDepth: 2,
+      onFocus: this.focusItem,
+      renderItem: (item, depth, focused, _, expanded, { setExpanded }) => {
+        const arrow = Arrow({
+          className: classnames("arrow",
+                                { expanded: expanded,
```

Comment:
Very popular library for conditionally adding class names: https://www.npmjs.com/package/classnames

Instead of using string concat, give it an object and it'll apply only the class names which are truthy.